File: laya_snake/scores.py
import json
import os

SCORES_PATH = os.path.join(os.path.expanduser("~"), ".laya_snake", "best_scores.json")


def _load_all():
    try:
        with open(SCORES_PATH) as f:
            return json.load(f)
    except (FileNotFoundError, json.JSONDecodeError, OSError):
        return {}


def get_best(width, height):
    return _load_all().get(f"{width}x{height}", 0)


def set_best(width, height, score):
    """Persists score as the new best for this board size if it's higher.
    Returns the (possibly unchanged) best on record."""
    data = _load_all()
    key = f"{width}x{height}"
    current = data.get(key, 0)
    if score <= current:
        return current
    data[key] = score
    try:
        os.makedirs(os.path.dirname(SCORES_PATH), exist_ok=True)
        with open(SCORES_PATH, "w") as f:
            json.dump(data, f, indent=2)
    except OSError:
        pass
    return score

File: src/scores.rs
use serde_json::Value;
use std::{fs, path::PathBuf};

fn scores_path() -> PathBuf {
    let home = std::env::var_os("HOME")
        .or_else(|| std::env::var_os("USERPROFILE"))
        .map(PathBuf::from)
        .unwrap_or_else(|| PathBuf::from("."));
    home.join(".laya_snake").join("best_scores.json")
}

fn load_all() -> serde_json::Map<String, Value> {
    let path = scores_path();
    match fs::read_to_string(path) {
        Ok(text) => serde_json::from_str(&text).unwrap_or_default(),
        Err(_) => serde_json::Map::new(),
    }
}

pub fn get_best(width: usize, height: usize) -> u32 {
    load_all()
        .get(&format!("{width}x{height}"))
        .and_then(Value::as_u64)
        .unwrap_or(0) as u32
}

pub fn set_best(width: usize, height: usize, score: u32) -> u32 {
    let mut data = load_all();
    let key = format!("{width}x{height}");
    let current = data.get(&key).and_then(Value::as_u64).unwrap_or(0) as u32;
    if score <= current {
        return current;
    }
    data.insert(key, Value::from(score));
    let path = scores_path();
    if let Some(parent) = path.parent() {
        let _ = fs::create_dir_all(parent);
    }
    if let Ok(json) = serde_json::to_string_pretty(&data) {
        let _ = fs::write(path, json);
    }
    score
}
